src/bin/util/softhsm2-util-botan.cpp

```diff
@@ -207,6 +210,53 @@ int crypto_import_key_pair
 	return result;
 }
 
+int crypto_import_certificate
+(
+	CK_SESSION_HANDLE hSession,
+	char* filePath,
+	char* label,
+	char* objID,
+	size_t objIDLen
+)
+{
+	Botan::X509_Certificate cert(filePath);
+	std::vector<uint8_t> blob = cert.BER_encode();
+	std::vector<uint8_t> name = cert.subject_dn().BER_encode();
+	std::vector<uint8_t> issuer = cert.issuer_dn().BER_encode();
+
+	std::vector<uint8_t> serial;
+	Botan::DER_Encoder der(serial);
+	der.encode(Botan::BigInt::decode(cert.serial_number()));
+
+	CK_OBJECT_CLASS certClass = CKO_CERTIFICATE;
+	CK_CERTIFICATE_TYPE certType = CKC_X_509;
+	CK_BBOOL ckTrue = CK_TRUE, ckFalse = CK_FALSE, ckToken = CK_TRUE;
+	CK_ATTRIBUTE certTemplate[] = {
+		{ CKA_CLASS,            &certClass,       sizeof(certClass) },
+		{ CKA_CERTIFICATE_TYPE, &certType,        sizeof(certType) },
+		{ CKA_LABEL,            label,            strlen(label) },
+		{ CKA_ID,               objID,            objIDLen },
+		{ CKA_TOKEN,            &ckToken,         sizeof(ckToken) },
+		{ CKA_PRIVATE,          &ckFalse,         sizeof(ckTrue) },
+		{ CKA_VALUE,            &*blob.begin(),   (CK_ULONG)blob.size() },
```

Comment:
It doesn't seem that this function is doing error checking on the user inputs and resulting vectors.

A possible result of that is that it will be dereferencing `begin()` of an empty vector, which is not allowed (https://cplusplus.com/reference/vector/vector/begin/) and would cause undefined behaviour.

Reply:
This helps, thanks for changing it so quickly.